### PR TITLE
Refactor window to class: TrackDesignPlace

### PR DESCRIPTION
**Pull request by 94794**
Should do #13811.

### CODE REVIEW

**Review by tupaschoal — COMMENTED** [review]
Hello @94794 and welcome to the project!
It seems like there is a merge conflict with one of the files you touched. Can you sort it out so we can get on with the reviewing?
If you need help, let us know here or over on our Discord server.

**Comment by 94794** [comment]
Sorry for the mess, accidentally merged the develop into the fix branch.

Should be ready now.

File: src/openrct2-ui/windows/TrackDesignPlace.cpp [hunk 1]
```diff
@@ -70,633 +70,553 @@ static rct_widget window_track_place_widgets[] = {
...
-                    pixel[168 + i] = colour; //        y + 1
-                    pixel[2 + i] = colour;   // x + 2
-                    pixel[172 + i] = colour; // x + 4, y + 1
-                }
-            }
-        }
+    auto* window = window_bring_to_front_by_class(WC_TRACK_DESIGN_PLACE);
+    if (window == nullptr)
+    {
+        window = WindowCreate<TrackDesignPlaceWindow>(WC_TRACK_DESIGN_PLACE, WW, WH, 0);
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/TrackDesignPlace.cpp:592]: There is a `WindowFocusOrCreate` that you can use to handle the window_bring_to_front_by_class all as a oner.

File: src/openrct2-ui/windows/TrackDesignPlace.cpp [hunk 2]
```diff
@@ -70,633 +70,553 @@ static rct_widget window_track_place_widgets[] = {
...
-}
+};
 
-static void WindowTrackPlaceDrawMiniPreviewMaze(
-    TrackDesign* td6, int32_t pass, const CoordsXY& origin, CoordsXY min, CoordsXY max)
+rct_window* WindowTrackPlaceOpen(const track_design_file_ref* tdFileRef)
 {
-    uint8_t rotation = (_currentTrackPieceDirection + get_current_rotation()) & 3;
-    for (const auto& mazeElement : td6->maze_elements)
+    _trackDesign = TrackDesignImport(tdFileRef->path);
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/TrackDesignPlace.cpp:579]: the trackdesign should become owned by the window. I suggest you add a public method to the window called `Initialise(std::unique_ptr<TrackDesign>&& trackDesign)` and you move the loaded track design into the window.

File: src/openrct2-ui/windows/TrackDesignPlace.cpp [hunk 3]
```diff
@@ -70,633 +70,553 @@ static rct_widget window_track_place_widgets[] = {
...
-/**
- *
- *  rct2: 0x006CFF01
- */
-static void WindowTrackPlaceUnknown14(rct_window* w)
-{
-    WindowTrackPlaceDrawMiniPreview(_trackDesign.get());
-}
-
-static void WindowTrackPlaceInvalidate(rct_window* w)
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/TrackDesignPlace.cpp:399]: This has gone missing. It should become `OnPrepareDraw`

File: src/openrct2-ui/windows/TrackDesignPlace.cpp [hunk 4]
```diff
@@ -70,633 +70,553 @@ static rct_widget window_track_place_widgets[] = {
...
 // clang-format on
 
 static std::vector<uint8_t> _window_track_place_mini_preview;
 static CoordsXY _windowTrackPlaceLast;
-
 static RideId _window_track_place_ride_index;
 static bool _window_track_place_last_was_valid;
-static CoordsXYZ _windowTrackPlaceLastValid;
 static money32 _window_track_place_last_cost;
+static CoordsXYZ _windowTrackPlaceLastValid;
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/TrackDesignPlace.cpp:80]: In general we are wanting to move all of these static variables into the window. You will need to modify the functions that use these directly from other places e.g. `TrackPlaceClearProvisionalTemporarily` and `TrackPlaceRestoreProvisional` make it so those functions find the track window with `window_find_by_class` and then call a public function.

```cpp
void TrackPlaceClearProvisionalTemporarily(){
    auto* trackPlaceWnd = static_cast<TrackDesignPlaceWindow*>(window_find_by_class(WC_TRACK_DESIGN_PLACE));
    if (trackPlaceWnd != nullptr){
        trackPlaceWnd->TrackPlaceClearProvisionalTemporarily();
    }
}

File: src/openrct2-ui/windows/TrackDesignPlace.cpp [hunk 5]
```diff
@@ -70,633 +70,553 @@ static rct_widget window_track_place_widgets[] = {
...
-    auto ft = Formatter::Common();
-    ft.Add<char*>(_trackDesign->name.c_str());
-    WindowDrawWidgets(w, dpi);
+        return z + TrackDesignGetZPlacement(_trackDesign.get(), GetOrAllocateRide(PreviewRideId), { loc, z });
+    }
 
-    // Draw mini tile preview
-    rct_drawpixelinfo clippedDpi;
-    if (clip_drawpixelinfo(&clippedDpi, dpi, w->windowPos + ScreenCoordsXY{ 4, 18 }, 168, 78))
+    void WindowTrackPlaceDrawMiniPreview(TrackDesign* td6)
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/TrackDesignPlace.cpp:359]: For all of the methods that are now part of the class you should drop the `WindowTrackPlace`

**Review by duncanspumpkin — CHANGES_REQUESTED** [review]
Sorry it took a while to get around to reviewing. This all looks fine but a few little changes to be made.

**Comment by 94794** [comment]
Should be ok now.

**Comment by duncanspumpkin** [comment]
I've squashed this down and rebased it so that its easier to understand.

**Comment by duncanspumpkin** [comment]
I had a little play about with this and it all seems to work.

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**